### PR TITLE
native: add Echidna changes

### LINKED ISSUE #3549: Designation event extension
## Is your feature request related to a problem? Please describe.
I'm always frustrated when we don't have protocol parity with C# node. https://github.com/neo-project/neo/pull/3397 has implemented an event extension.

## Describe the solution you'd like
Port it to NeoGo.

### LINKED ISSUE #3550: Base64URL support
## Is your feature request related to a problem? Please describe.
I'm always frustrated when we're not on par with C# in terms of protocol development.  https://github.com/neo-project/neo/pull/3453 has extended Stdlib.

## Describe the solution you'd like
Port it to NeoGo.

### PR DESCRIPTION
**Pull request by AliceInHunterland**
Close #3549
Close #3550
Close https://github.com/nspcc-dev/neo-go/issues/3702

### CODE REVIEW

File: pkg/core/native/designate.go [hunk 1]
```diff
@@ -412,10 +420,27 @@ func (s *Designate) DesignateAsRole(ic *interop.Context, r noderoles.Role, pubs
...
+			stackitem.NewBigInteger(big.NewInt(int64(r))),
+			stackitem.NewBigInteger(big.NewInt(int64(ic.Block.Index))),
+			pubsToArray(oldNodes),
+			pubsToArray(newNodes),
+		}))
+	} else {
+		ic.AddNotification(s.Hash, DesignationEventName, stackitem.NewArray([]stackitem.Item{
+			stackitem.NewBigInteger(big.NewInt(int64(r))),
+			stackitem.NewBigInteger(big.NewInt(int64(ic.Block.Index))),
+		}))
```
Comment (AnnaShaleva) [inline pkg/core/native/designate.go:442]: It's a common code, it needs to be unified.

File: pkg/core/native/designate.go [hunk 2]
```diff
@@ -412,10 +420,27 @@ func (s *Designate) DesignateAsRole(ic *interop.Context, r noderoles.Role, pubs
...
-	}))
+	if ic.IsHardforkEnabled(config.HFEchidna) {
+		oldNodes, _, err := s.GetDesignatedByRole(ic.DAO, r, ic.Block.Index)
+		if err != nil {
+			return fmt.Errorf("failed to get old nodes for role %d: %w", r, err)
+		}
+		newNodes, _, err := s.GetDesignatedByRole(ic.DAO, r, ic.Block.Index+1)
+		if err != nil {
+			return fmt.Errorf("failed to get new nodes for role %d: %w", r, err)
+		}
```
Comment (AnnaShaleva) [inline pkg/core/native/designate.go:431]: It's inefficient, it will lead to one more storage seek, whereas you already have the new list of nodes available here.

File: pkg/core/native/std.go [hunk 3]
```diff
@@ -100,6 +100,15 @@ func newStd() *Std {
...
+	desc = newDescriptor("base64UrlEncode", smartcontract.StringType,
+		manifest.NewParameter("data", smartcontract.ByteArrayType))
+	md = newMethodAndPrice(s.base64UrlEncode, 1<<5, callflag.NoneFlag, config.HFEchidna)
+	s.AddMethod(md, desc)
+
+	desc = newDescriptor("base64UrlDecode", smartcontract.ByteArrayType,
+		manifest.NewParameter("s", smartcontract.StringType))
+	md = newMethodAndPrice(s.base64UrlDecode, 1<<5, callflag.NoneFlag, config.HFEchidna)
+	s.AddMethod(md, desc)
 	desc = newDescriptor("base58Encode", smartcontract.StringType,
```
Comment (AnnaShaleva) [inline pkg/core/native/std.go:112]: New line is needed, follow the style.

File: pkg/core/native/native_test/management_test.go [hunk 4]
```diff
@@ -56,13 +56,21 @@ var (
...
...airing","offset":28,"parameters":[{"name":"g1","type":"InteropInterface"},{"name":"g2","type":"InteropInterface"}],"returntype":"InteropInterface","safe":true},{"name":"bls12381Serialize","offset":35,"parameters":[{"name":"g","type":"InteropInterface"}],"returntype":"ByteArray","safe":true},{"name":"keccak256","offset":42,"parameters":[{"name":"data","type":"ByteArray"}],"returntype":"ByteArray","safe":true},{"name":"murmur32","offset":49,"parameters":[{"name":"data","type":"ByteArray"},{"name":"seed","type":"Integer"}],"returntype":"ByteArray","safe":true},{"name":"ripemd160","offset":56,"parameters":[{"name":"data","type":"ByteArray"}],"returntype":"ByteArray","safe":true},{"name":"sha256","offset":63,"parameters":[{"name":"data","type":"ByteArray"}],"returntype":"ByteArray","safe":true},{"name":"verifyWithECDsa","offset":70,"parameters":[{"name":"message","type":"ByteArray"},{"name":"pubkey","type":"ByteArray"},{"name":"signature","type":"ByteArray"},{"name":"curveHash","type":"Integer"}],"returntype":"Boolean","safe":true}],"events":[]},"features":{},"groups":[],"permissions":[{"contract":"*","methods":"*"}],"supportedstandards":[],"trusts":[],"extra":null},"updatecounter":0}`,
```
[the remark this answers is not in the document]
Comment (AliceInHunterland) [reply pkg/core/native/native_test/management_test.go:61]: Script and checksum differ from cockatriceCSS CryptoLib. I do not understand why.

**Comment by AnnaShaleva** [comment]
@AliceInHunterland ready for review? Tests are still failing.

File: pkg/core/native/designate.go [hunk 5]
```diff
@@ -412,10 +420,20 @@ func (s *Designate) DesignateAsRole(ic *interop.Context, r noderoles.Role, pubs
 		return fmt.Errorf("failed to update Designation role data cache: %w", err)
 	}
 
-	ic.AddNotification(s.Hash, DesignationEventName, stackitem.NewArray([]stackitem.Item{
+	notification := stackitem.NewArray([]stackitem.Item{
```
Comment (AnnaShaleva) [inline pkg/core/native/designate.go:423]: s/notification/ntf ?

I saw it several times before in previous PRs, let's agree to use short names for local function variables (if it's possible and if the variable meaning can easily be guessed from the short name). Long names are suitable for exported functions, exported package variables/constants and etc. to give some more context to the user.

File: pkg/core/native/native_test/common_test.go [hunk 6]
```diff
@@ -117,20 +117,25 @@ func testGetSetCache(t *testing.T, c *neotest.ContractInvoker, name string, defa
...
 	if ok {
 		h := designateInvoker.Invoke(t, stackitem.Null{}, "designateAsRole", int64(r), pubs)
+		notification := stackitem.NewArray([]stackitem.Item{
+			stackitem.Make(int64(r)),
+			stackitem.Make(designateInvoker.Chain.BlockHeight()),
+		})
+		if len(oldKeys) > 0 {
+			notification.Append(stackitem.NewArray([]stackitem.Item{stackitem.NewByteArray(oldKeys[0][0].Bytes())}))
+			notification.Append(stackitem.NewArray([]stackitem.Item{stackitem.NewByteArray(nodes[0].Bytes())}))
+		}
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/common_test.go:134]: This way the test is not properly constructed. You need to take not only `[0]`-th element of nodes/oldKeys.

File: pkg/core/native/native_test/common_test.go [hunk 7]
```diff
@@ -117,20 +117,25 @@ func testGetSetCache(t *testing.T, c *neotest.ContractInvoker, name string, defa
...
 		designateInvoker.CheckTxNotificationEvent(t, h, 0, state.NotificationEvent{
 			ScriptHash: designateInvoker.Hash,
 			Name:       native.DesignationEventName,
-			Item: stackitem.NewArray([]stackitem.Item{
-				stackitem.Make(int64(r)),
-				stackitem.Make(designateInvoker.Chain.BlockHeight()),
-			}),
+			Item:       notification,
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/common_test.go:138]: Ditto for naming.

File: pkg/core/native/native_test/designate_test.go [hunk 8]
```diff
@@ -56,6 +56,29 @@ func TestDesignate_DesignateAsRole(t *testing.T) {
 	})
 }
 
+func TestDesignate_DesignateAsRoleHF(t *testing.T) {
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/designate_test.go:59]: > TestDesignate_DesignateAsRoleHF

TestDesignate_DesignateAsRole_Echidna

We have a lot of hardforks.

File: pkg/core/native/native_test/designate_test.go [hunk 9]
```diff
@@ -56,6 +56,29 @@ func TestDesignate_DesignateAsRole(t *testing.T) {
...
+	})
+	designateInvoker := c.WithSigners(c.Committee)
+
+	priv, err := keys.NewPrivateKey()
+	require.NoError(t, err)
+	oldPubs := keys.PublicKeys{priv.PublicKey()}
+
+	checkNodeRoles(t, designateInvoker, true, noderoles.Oracle, 0, keys.PublicKeys{})                            //BlockHeight is 1.
+	setNodesByRole(t, designateInvoker, true, noderoles.Oracle, oldPubs)                                         //BlockHeight is 2.
+	checkNodeRoles(t, designateInvoker, true, noderoles.Oracle, designateInvoker.Chain.BlockHeight()+1, oldPubs) //BlockHeight is 3.
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/designate_test.go:73]: Missing space in the start of the comment, here and below.

File: pkg/core/native/native_test/management_test.go [hunk 10]
```diff
@@ -56,13 +56,20 @@ var (
...
...ntype":"Boolean","safe":false},{"name":"unclaimedGas","offset":119,"parameters":[{"name":"account","type":"Hash160"},{"name":"end","type":"Integer"}],"returntype":"Integer","safe":true},{"name":"unregisterCandidate","offset":126,"parameters":[{"name":"pubkey","type":"PublicKey"}],"returntype":"Boolean","safe":false},{"name":"vote","offset":133,"parameters":[{"name":"account","type":"Hash160"},{"name":"voteTo","type":"PublicKey"}],"returntype":"Boolean","safe":false}],"events":[{"name":"Transfer","parameters":[{"name":"from","type":"Hash160"},{"name":"to","type":"Hash160"},{"name":"amount","type":"Integer"}]},{"name":"CandidateStateChanged","parameters":[{"name":"pubkey","type":"PublicKey"},{"name":"registered","type":"Boolean"},{"name":"votes","type":"Integer"}]},{"name":"Vote","parameters":[{"name":"account","type":"Hash160"},{"name":"from","type":"PublicKey"},{"name":"to","type":"PublicKey"},{"name":"amount","type":"Integer"}]},{"name":"CommitteeChanged","parameters":[{"name":"old","type":"Array"},{"name":"new","type":"Array"}]}]},"features":{},"groups":[],"permissions":[{"contract":"*","methods":"*"}],"supportedstandards":["NEP-17"],"trusts":[],"extra":null},"updatecounter":0}`,
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/management_test.go:61]: Why do you have Neo here? It shouldn't be here, only Designation is changed.

File: pkg/core/native/native_test/management_test.go [hunk 11]
```diff
@@ -56,13 +56,21 @@ var (
...
 )
 
 func init() {
 	for k, v := range defaultCSS {
 		if _, ok := cockatriceCSS[k]; !ok {
 			cockatriceCSS[k] = v
 		}
+		if _, ok := echidnaCSS[k]; !ok {
+			echidnaCSS[k] = v
+		}
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/native/native_test/management_test.go:73]: You should firstly properly fill `cockatriceCSS` (since it's a map) and only after this in a new cycle properly fill Echidna's states.

File: pkg/core/native/designate.go [hunk 12]
```diff
@@ -121,7 +121,15 @@ func newDesignate(initialNodeRoles map[noderoles.Role]keys.PublicKeys) *Designat
 	eDesc := newEventDescriptor(DesignationEventName,
 		manifest.NewParameter("Role", smartcontract.IntegerType),
 		manifest.NewParameter("BlockIndex", smartcontract.IntegerType))
-	eMD := newEvent(eDesc)
+	eMD := newEvent(eDesc, config.HFDefault, config.HFEchidna)
```
Comment (AnnaShaleva) [inline pkg/core/native/designate.go:124]: >  native: add Designation event extension after HFEchidna 

Why `after`? It's valid starting from Echidna, so it's either `to` or `starting from`.

File: pkg/core/native/std_test.go [hunk 13]
```diff
@@ -170,7 +170,9 @@ func TestStdLibJSON(t *testing.T) {
 func TestStdLibEncodeDecode(t *testing.T) {
 	s := newStd()
 	original := []byte("my pretty string")
+	originalUrl := []byte("Subject=test@example.com&Issuer=https://example.com")
 	encoded64 := base64.StdEncoding.EncodeToString(original)
+	encoded64Url := "U3ViamVjdD10ZXN0QGV4YW1wbGUuY29tJklzc3Vlcj1odHRwczovL2V4YW1wbGUuY29t"
```
Comment (AnnaShaleva) [inline pkg/core/native/std_test.go:175]: Add a comment explaining where did you get these data.

Comment (AnnaShaleva) [reply pkg/core/native/std_test.go:175]: It's needed because if something will be changed, we need to preserve compatibility with C#.

File: pkg/core/native/native_neo.go [hunk 14]
```diff
@@ -192,18 +192,27 @@ func newNEO(cfg config.ProtocolConfiguration) *NEO {
 
 	desc = newDescriptor("registerCandidate", smartcontract.BoolType,
 		manifest.NewParameter("pubkey", smartcontract.PublicKeyType))
-	md = newMethodAndPrice(n.registerCandidate, 0, callflag.States)
+	md = newMethodAndPrice(n.registerCandidate, 0, callflag.States, config.HFDefault, config.HFEchidna)
+	n.AddMethod(md, desc)
+
+	md = newMethodAndPrice(n.registerCandidate, 0, callflag.States|callflag.AllowNotify, config.HFEchidna)
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:198]: > native: support callflags-based native method after HFEchidna 

Turns out that callflag-based native forks are already supported. So the commit doesn't add anything new wrt this functionality. The commit itself does something different, adjust the message accordingly.

File: examples/engine/go.mod [hunk 15]
```diff
@@ -2,4 +2,4 @@ module github.com/nspcc-dev/neo-go/examples/engine
 
 go 1.22
 
-require github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241212130705-ea0a6114d2d6
+require github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241219193620-20b74b038f23
```
Comment (AnnaShaleva) [inline examples/engine/go.mod:5]: > interop: support callflags-based native method after HFEchidna 

https://github.com/nspcc-dev/neo-go/pull/3155/commits/c6d259b28577725a971987157c0e0be37dd763da

File: examples/nft-nd-nns/go.mod [hunk 16]
```diff
@@ -4,7 +4,7 @@ go 1.22
 
 require (
 	github.com/nspcc-dev/neo-go v0.106.4-0.20241016130346-d8e945978af6
-	github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241212130705-ea0a6114d2d6
+	github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241219193620-20b74b038f23
```
Comment (AnnaShaleva) [inline examples/nft-nd-nns/go.mod:7]: neo-go dependency should be updated afterwords in a separate commit.

File: pkg/core/native/native_neo.go (hunk 14 above)
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:198]: BTW, each extension should be properly documented in `Hardforks` section of `./docs/node-configuration.md`.

File: examples/nft-nd-nns/go.mod (hunk 16 above)
Comment (AliceInHunterland) [reply examples/nft-nd-nns/go.mod]: similar to https://github.com/nspcc-dev/neo-go/commit/c6d259b28577725a971987157c0e0be37dd763da ? My `./scripts/update_deps.sh 69c75acb126038a719d8a6ed3097fb20a11d50d1` update all of them and if update only internal dependencies then Tests / Check internal dependencies (pull_request) fails.

Comment (AnnaShaleva) [reply examples/nft-nd-nns/go.mod]: > similar to https://github.com/nspcc-dev/neo-go/commit/c6d259b28577725a971987157c0e0be37dd763da ?

No. This commit updates `interop` dependencies. What you need is for `nft-nd-nns` manually run `go get github.com/nspcc-dev/neo-go@...` with latest commit in your PR.

**Comment by AnnaShaleva** [comment]
@AliceInHunterland ready for review?

File: examples/nft-nd-nns/go.mod (hunk 16 above)
Comment (AliceInHunterland) [reply examples/nft-nd-nns/go.mod]: its not okay that the test is still failing?

File: examples/nft-nd-nns/go.sum [hunk 17]
```diff
@@ -78,6 +78,8 @@ github.com/nspcc-dev/neo-go v0.106.4-0.20241016130346-d8e945978af6 h1:3/V1U2ZgbF
 github.com/nspcc-dev/neo-go v0.106.4-0.20241016130346-d8e945978af6/go.mod h1:ds91T4WJwtk7eWUo0fuVC36HpTQKkkdj5AjNxbjXAR0=
 github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241212130705-ea0a6114d2d6 h1:rTnsU+Y/bP1bLN/SNWmOKEexmSeniMQe5bOJxXNbXgg=
 github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241212130705-ea0a6114d2d6/go.mod h1:kVLzmbeJJdbIPF2bUYhD8YppIiLXnRQj5yqNZvzbOL0=
+github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241223073935-69c75acb1260 h1:8dlMzyuJGghY1NEIXFr3wWJ5ImgjiZgi+0OWRBruU90=
+github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241223073935-69c75acb1260/go.mod h1:kVLzmbeJJdbIPF2bUYhD8YppIiLXnRQj5yqNZvzbOL0=
```
Comment (AnnaShaleva) [inline examples/nft-nd-nns/go.sum:82]: Something is wrong. The rest of `go.mod` files must be updated.

**Review by AnnaShaleva — COMMENTED** [review]
Also need `neo-go` dependency upgrade.

File: examples/nft-nd-nns/go.mod (hunk 16 above)
Comment (AnnaShaleva) [reply examples/nft-nd-nns/go.mod]: No. You firstly need to execute `scripts/update-deps.sh`, then commit changes, then push them into master, then manually upgrade `neo-go` dependency for NNS.

File: pkg/core/native/designate.go [hunk 18]
```diff
@@ -412,10 +420,20 @@ func (s *Designate) DesignateAsRole(ic *interop.Context, r noderoles.Role, pubs
 		return fmt.Errorf("failed to update Designation role data cache: %w", err)
 	}
 
-	ic.AddNotification(s.Hash, DesignationEventName, stackitem.NewArray([]stackitem.Item{
+	ntf := stackitem.NewArray([]stackitem.Item{
 		stackitem.NewBigInteger(big.NewInt(int64(r))),
 		stackitem.NewBigInteger(big.NewInt(int64(ic.Block.Index))),
-	}))
+	})
+
```
Comment (AnnaShaleva) [inline pkg/core/native/designate.go:427]: Useless line.

File: pkg/core/native/native_test/common_test.go [hunk 19]
```diff
@@ -117,20 +117,33 @@ func testGetSetCache(t *testing.T, c *neotest.ContractInvoker, name string, defa
...
+			old := stackitem.NewArray([]stackitem.Item{})
+			for _, key := range oldKeys[0] {
+				old.Append(stackitem.NewByteArray(key.Bytes()))
+			}
+			newKeys := stackitem.NewArray([]stackitem.Item{})
+			for _, key := range nodes {
+				newKeys.Append(stackitem.NewByteArray(key.Bytes()))
+			}
+			ntf.Append(old)
+			ntf.Append(newKeys)
```
Comment (AnnaShaleva) [inline pkg/core/native/native_test/common_test.go:141]: Good, now it's correct.

File: examples/nft-nd-nns/go.mod [hunk 20]
```diff
@@ -4,7 +4,7 @@ go 1.22
 
 require (
 	github.com/nspcc-dev/neo-go v0.106.4-0.20241016130346-d8e945978af6
-	github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241212130705-ea0a6114d2d6
+	github.com/nspcc-dev/neo-go/pkg/interop v0.0.0-20241223113753-f3e22b707014
```
Comment (AnnaShaleva) [inline examples/nft-nd-nns/go.mod:7]: Good, and then in a separate commit you need to update `github.com/nspcc-dev/neo-go` dependency.

**Review by AnnaShaleva — APPROVED** [review]
